File: src/go/problems/MajorityElement_test.go
// https://leetcode.com/problems/majority-element/

package problems

import (
	"testing"
)

func TestMajorityElement(t *testing.T) {
	result := MajorityElement([]int{2, 2, 11, 2})
	t.Log(result)
}

func MajorityElement(nums []int) int {
	result := 0
	count := 0

	for _, num := range nums {
		if count == 0 {
			result = num
		}

		if result == num {
			count++
		} else {
			count--
		}
	}

	return result
}

File: src/python/problems/majority_element_test.py
#https://leetcode.com/problems/majority-element/
from core.problem_base import *

class MajorityElement(ProblemBase):
    def Solution(self, nums: list[int]) -> int:
        result = 0
        count = 0

        for num in nums:
            if count == 0:
                result = num

            count += 1 if result == num else -1
        
        return result


if __name__ == '__main__':
    TestGen(MajorityElement) \
        .Add(lambda tc: tc.Param([3,2,3]).Result(3)) \
        .Add(lambda tc: tc.Param([2,2,1,1,1,2,2]).Result(2)) \
        .Run()
